### DISCUSSION TITLE
I need to enable Fortran support during PETSc installation

### DISCUSSION THREAD

**Discussion by ddmoa** [idaholab/moose#29561] [Q&A Getting Started]
### Check these boxes if you have followed the posting rules.

- [X] I have consulted the [Posting Guidelines](https://github.com/idaholab/moose/discussions/18270).
- [X] I have searched the [Discussions Forum](https://github.com/idaholab/moose/discussions) and [MOOSE Framework Troubleshooting](https://mooseframework.inl.gov/help/troubleshooting.html) and have not found what I was looking for
- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)

### Issue or question about MOOSE

Hello! I want to install PETSc with Fortran support when running the Moose installation script. I modified the configure_petsc.sh file by setting --with-fortran-bindings=1, but it doesn't seem to work. The final PETSc installation directory does not contain the petsc/finclude directory. How should I modify the configuration?

**Comment by lindsayad** [comment]
Either delete the line `--with-sowing=0` or change it to `--with-sowing=1`

**Reply by ddmoa** [reply]
I followed your suggestions, but it still isn't working. What confuses me is that the `petsc/arc-moose/include` directory in MOOSE doesn't contain header files like `petsc.h` or `petscao.h` , even though these files are present in the standalone PETSc installation.

I copied all these header files directly into the MOOSE directory, but I'm not sure if this will work. Could you give me some advice?

**Reply by milljm** [reply]
The following does not seem to create an `finclude` directory either:

```bash
git clone -b release https://gitlab.com/petsc/petsc.git petsc
cd petsc/
git co v3.22.1
./configure --with-64-bit-indices \
      --with-cxx-dialect=C++17 \
      --with-debugging=no \
      --with-fortran-bindings=1 \
      --with-mpi=1 \
      --with-openmp=1 \
      --with-strict-petscerrorcode=1 \
      --with-shared-libraries=1 \
      --with-sowing=1 \
      --download-fblaslapack=1 \
      --download-hpddm=1 \
      --download-hypre=1 \
      --download-metis=1 \
      --download-mumps=1 \
      --download-ptscotch=1 \
      --download-parmetis=1 \
      --download-scalapack=1 \
      --download-slepc=1 \
      --download-strumpack=1 \
      --download-superlu_dist=1 \
      --prefix=/data/milljm/petsc_installed
make PETSC_DIR=/data/milljm/projects/petsc PETSC_ARCH=arch-linux-c-opt all
make PETSC_DIR=/data/milljm/projects/petsc PETSC_ARCH=arch-linux-c-opt install
```
What is supposed to happen? This is a clone directly from PETSc. No use of MOOSE at all here.

**Reply by lindsayad** [reply]
> doesn't contain header files like `petsc.h` or `petscao.h`

These files are in `$PETSC_DIR/include`, not `$PETSC_DIR/$PETSC_ARCH/include`. When building with a `PETSC_ARCH` like is done when running the `update_and_rebuild_petsc.sh` script, you will not see all the files arranged such as you would see for a classic installation. C headers independent of configuration will be in `$PETSC_DIR/include` (`moose/petsc/include`) while headers dependent on configuration will be in `$PETSC_DIR/$PETSC_ARCH/include` (`moose/petsc/arch-moose/include`)

**Reply by ddmoa** [reply]
Sorry, I don't understand how the configuration-dependent headers and configuration-independent headers are set.

Maybe I need to clarify the points I mentioned above about the headers. I downloaded PETSc version 3.21.5 and configured it using only the following command: `./configure --prefix=/xx/build --with-mpi-dir=/xx/build --download-fblaslapack=1` After installation, the `build` directory contains a `include/petsc/finclude` directory, and inside it, there are headers like `petsc.h`.

**Reply by lindsayad** [reply]
that is definitely different from what I expected based on your OP from which I thought you only modified the fortran configuration in `configure_petsc.sh`. So you are running your configure command in the PETSc root directory?

**Reply by ddmoa** [reply]
Yes, I ran the command directly in the root directory. I don't think there's any fundamental difference between this and using the script.

**Reply by lindsayad** [reply]
If you had been running our script then you would have been brining along a lot of configure arguments, so that's a substantial difference, and we do not use `--prefix` which is another substantial difference

**Reply by lindsayad** [reply]
I don't understand at this point what's not working. When running your own configure command you are not getting fortran bindings?

**Reply by ddmoa** [reply]
It seems that the `--prefix` option is what makes the difference. For the PETSc I installed myself, if I don't use the `--prefix` option to set a custom installation path, those header files I mentioned earlier do not exist. However, if I use `--prefix` to set the installation directory to build, then `build/include` will contain `petsc/finclude`.

If I add the `--prefix` option to the PETSc installation script in MOOSE and specify another location, will MOOSE still work properly?

**Reply by lindsayad** [reply]
Yes. You just will need to set `PETSC_DIR` to that location and make sure `PETSC_ARCH` is not set

**Reply by ddmoa** [reply]
So, do I need to remove the `PETSC_ARCH` setting in the MOOSE script? Additionally, it seems that the `PETSC_DIR` path should be `moose/petsc`, but the path I set using `--prefix` is `moose/petsc/build`. Should I change the `PETSC_DIR` in the script to `moose/petsc/build` instead?

**Reply by lindsayad** [reply]
You can control the prefix destination with the environment variable `PETSC_PREFIX`. You should not need to do anything about `PETSC_DIR`. I would remove the setting of `PETSC_ARCH` in the script, yes

**Reply by ddmoa** [reply]
Okay, after setting `PETSC_PREFIX`, two directories are created: one is the custom installation directory (`build`) that I set, and the other is the default directory. After installation, the various libraries are distributed between these two locations. It seems that some of the dependencies MOOSE requires were installed in the `build` directory.

When I tested it, I got the following error message:

```Linking Library /home/hpc/projects/moose/framework/libmoose-opt.la...
/usr/bin/grep: /home/hpc/projects/moose/petsc/arch-moose/lib/libhdf5_hl.la: No such file or directory
/usr/bin/sed: can't read /home/hpc/projects/moose/petsc/arch-moose/lib/libhdf5_hl.la: No such file or directory
libtool:   error: '/home/hpc/projects/moose/petsc/arch-moose/lib/libhdf5_hl.la' is not a valid libtool archive
make: *** [/home/hpc/projects/moose/framework/moose.mk:423: /home/hpc/projects/moose/framework/libmoose-opt.la] Error 1
make: *** Waiting for unfinished jobs....
```

It looks like `libhdf5_hl` was installed in `build/lib`. How can I resolve this issue?

**Reply by lindsayad** [reply]
That's silly. I would have to spend some time investigating to figure out what's happening there. Is there a reason you can't just use your "custom" PETSc install? It sounds like that worked for you so do you need our script?
